pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java

```diff
@@ -49,57 +47,67 @@
           RelDistribution.Type.BROADCAST_DISTRIBUTED, RelDistribution.Type.HASH_DISTRIBUTED);
-  private static final Random RANDOM = new Random();
-  // TODO: Deduct this value via grpc config maximum byte size; and make it configurable with override.
-  // TODO: Max block size is a soft limit. only counts fixedSize datatable byte buffer
-  private static final int MAX_MAILBOX_CONTENT_SIZE_BYTES = 4 * 1024 * 1024;
-
-  private final List<ServerInstance> _receivingStageInstances;
-  private final RelDistribution.Type _exchangeType;
-  private final KeySelector<Object[], Object[]> _keySelector;
-  private final String _serverHostName;
-  private final int _serverPort;
-  private final long _jobId;
-  private final int _stageId;
-  private final MailboxService<TransferableBlock> _mailboxService;
-  private final DataSchema _dataSchema;
-  private final boolean _isLeafStageSender;
-  private Operator<TransferableBlock> _dataTableBlockBaseOperator;
-
-  public MailboxSendOperator(MailboxService<TransferableBlock> mailboxService, DataSchema dataSchema,
+
+  private final Operator<TransferableBlock> _dataTableBlockBaseOperator;
+  private final BlockExchange _exchange;
+
+  @VisibleForTesting
+  interface BlockExchangeFactory {
+    BlockExchange build(MailboxService<TransferableBlock> mailboxService, List<MailboxIdentifier> destinations,
+        RelDistribution.Type exchange, KeySelector<Object[], Object[]> selector, BlockSplitter splitter);
+  }
+
+  @VisibleForTesting
+  interface MailboxIdGenerator {
+    MailboxIdentifier generate(ServerInstance server);
+  }
+
+  public MailboxSendOperator(MailboxService<TransferableBlock> mailboxService,
       Operator<TransferableBlock> dataTableBlockBaseOperator, List<ServerInstance> receivingStageInstances,
       RelDistribution.Type exchangeType, KeySelector<Object[], Object[]> keySelector, String hostName, int port,
       long jobId, int stageId, boolean isLeafStageSender) {
-    _dataSchema = dataSchema;
-    _mailboxService = mailboxService;
+    this(mailboxService, dataTableBlockBaseOperator, receivingStageInstances, exchangeType, keySelector,
+        isLeafStageSender, server -> toMailboxId(server, jobId, stageId, hostName, port), BlockExchange::getExchange);
+  }
+
+  @VisibleForTesting
+  MailboxSendOperator(MailboxService<TransferableBlock> mailboxService,
+      Operator<TransferableBlock> dataTableBlockBaseOperator, List<ServerInstance> receivingStageInstances,
+      RelDistribution.Type exchangeType, KeySelector<Object[], Object[]> keySelector,
+      boolean isLeafStageSender, MailboxIdGenerator mailboxIdGenerator, BlockExchangeFactory blockExchangeFactory) {
     _dataTableBlockBaseOperator = dataTableBlockBaseOperator;
-    _exchangeType = exchangeType;
-    if (_exchangeType == RelDistribution.Type.SINGLETON) {
+
+    List<MailboxIdentifier> receivingMailboxes;
+    if (exchangeType == RelDistribution.Type.SINGLETON) {
+      // TODO: this logic should be moved into SingletonExchange
```

Comment:
any reason we can't create a SingletonExchange? in this PR? i won't imagine too much code change right?

Reply:
I did create a `SingletonExchange` - this piece of code was confusing to me though... I wasn't really sure what it was doing and I didn't want to mess with it since there's no existing code coverage. Basically I was trying to understand why we're even iterating at all throughout the server instances instead of just throwing an exception if there's more than one?

Reply:
good question. i guess it requires a separate clean up.